text/0000-default-and-expanded-rustc-errors.md

```diff
@@ -0,0 +1,290 @@
+
+```
+error: cannot move out of borrowed content
+   --> /Users/jturner/Source/errors/borrowck-move-out-of-vec-tail.rs:30:17
+
+I’m trying to track the ownership of the contents of `tail`, which is borrowed, through this match statement:
+
+29  |>             match tail {
+
+In this match, you use an expression of the form [...].  When you do this, it’s like you are opening up the `tail` 
+value and taking out its contents. Because `tail` is borrowed, you can’t safely move the contents.
+
+30  |>                 [Foo { string: aa },
+    |>                                ^^ cannot move out of borrowed content
+
+You can avoid moving the contents out by working with each part using a reference rather than a move.  A naive fix 
+might look this:
+
+30  |>                 [Foo { string: ref aa },
+ 
+```
+
+# Detailed design
+
+The RFC is separated into two parts: the format of error messages and the format of expanded error messages (using `--explain errors`).
+
+## Format of error messages
+
+The proposal is a lighter error format focused on the code the user wrote.  Messages that help understand why an error occurred appear as labels on the source.  The goals of this new format are to:
+
+* Create something that's visually easy to parse
+* Remove noise/unnecessary information
+* Present information in a way that works well for new developers, post-onboarding, and experienced developers without special configuration
+* Draw inspiration from Elm as well as Dybuk and other systems that have already improved on the kind of errors that Rust has.
+
+In order to accomplish this, the proposed design needs to satisfy a number of constraints to make the result maximally flexible across various terminals:
+
+* Multiple errors beside each other should be clearly separate and not muddled together.
+* Each error message should draw the eye to where the error occurs with sufficient context to understand why the error occurs.
+* Each error should have a "header" section that is visually distinct from the code section.
+* Code should visually stand out from text and other error messages.  This allows the developer to immediately recognize their code.
+* Error messages should be just as readable when not using colors (eg for users of black-and-white terminals, color-impaired readers, weird color schemes that we can't predict, or just people that turn colors off)
+* Be careful using “ascii art” and avoid unicode. Instead look for ways to show the information concisely that will work across the broadest number of terminals.  We expect IDEs to possibly allow for a more graphical error in the future.
+* Where possible, use labels on the source itself rather than sentence "notes" at the end.  
+* Keep filename:line easy to spot for people who use editors that let them click on errors
+
+### Header
+
+```
+error[E0499]: cannot borrow `foo.bar1` as mutable more than once at a time
+  --> src/test/compile-fail/borrowck/borrowck-borrow-from-owned-ptr.rs:29:22
```

Comment:
Originally, the indent aligns with the column below it.  For example:

```
 --> file
  |
2 |
```
or
```
    --> file
     |
9999 |
```
IMHO, having some indent there still looks better and makes it easier to find the filename/line info, though there are other alternatives we could explore, like:

```
   --> file
     |
9999 |
```